File: src/backend/file-utils.js
const fs = require('fs-extra');
const path = require('path');
const os = require('os');

class FileUtils {
  constructor() {
    this.homeDir = os.homedir();
    this.sshManagerDir = path.join(this.homeDir, 'ssh_manager');
    this.sshDir = path.join(this.homeDir, '.ssh');
    this.sshConfigPath = path.join(this.sshDir, 'config');
  }

  // Nested Groups - Path Parsing Utilities
  
  /**
   * Parse a group path into its components
   * @param {string} groupPath - Group path like 'work/company-a/dev'
   * @returns {object} - { segments: ['work', 'company-a', 'dev'], depth: 3, parent: 'work/company-a', name: 'dev' }
   */
  parseGroupPath(groupPath) {
    if (!groupPath || typeof groupPath !== 'string') {
      throw new Error('Group path must be a non-empty string');
    }

    // Normalize path - remove leading/trailing slashes, collapse multiple slashes
    const normalized = groupPath.replace(/^\/+|\/+$/g, '').replace(/\/+/g, '/');
    
    if (!normalized) {
      throw new Error('Group path cannot be empty after normalization');
    }

    const segments = normalized.split('/');
    const depth = segments.length;
    const name = segments[segments.length - 1];
    const parent = depth > 1 ? segments.slice(0, -1).join('/') : null;

    return {
      original: groupPath,
      normalized,
      segments,
      depth,
      name,
      parent
    };
  }

  /**
   * Get all parent paths for a given group path
   * @param {string} groupPath - Group path like 'work/company-a/dev'
   * @returns {string[]} - ['work', 'work/company-a']
   */
  getParentPaths(groupPath) {
    const parsed = this.parseGroupPath(groupPath);
    const parents = [];
    
    for (let i = 1; i < parsed.segments.length; i++) {
      parents.push(parsed.segments.slice(0, i).join('/'));
    }
    
    return parents;
  }

  /**
   * Check if a group path is valid
   * @param {string} groupPath - Group path to validate
   * @returns {boolean} - Whether the path is valid
   */
  isValidGroupPath(groupPath) {
    try {
      const parsed = this.parseGroupPath(groupPath);
      
      // Each segment must be valid
      for (const segment of parsed.segments) {
        if (!this.isValidGroupName(segment)) {
          return false;
        }
      }
      
      // Additional nested path validations
      if (parsed.depth > 10) { // Reasonable nesting limit
        return false;
      }
      
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Check if a single group name (segment) is valid
   * @param {string} name - Single group name segment
   * @returns {boolean} - Whether the segment is valid
   */
  isValidGroupName(name) {
    if (!name || typeof name !== 'string') {
      return false;
    }
    
    // Allow letters, numbers, hyphens, underscores (but not slashes in individual segments)
    const validPattern = /^[a-z0-9][a-z0-9_-]*$/i;
    
    // Must be at least 2 characters and not too long
    return name.length >= 2 && name.length <= 50 && validPattern.test(name);
  }

  async ensureDirectoryStructure() {
    const directories = [
      this.sshManagerDir,
      path.join(this.sshManagerDir, 'config'),
      path.join(this.sshManagerDir, 'keys'),
      path.join(this.sshManagerDir, 'templates'),
      path.join(this.sshManagerDir, 'backups'),
      path.join(this.sshManagerDir, 'backups', 'config-backups')
    ];

    for (const dir of directories) {
      await fs.ensureDir(dir);
    }

    await fs.ensureDir(this.sshDir);

    const defaultGroups = ['work', 'personal', 'projects'];
    for (const group of defaultGroups) {
      await this.createGroup(group);
    }
  }

  async writeConfigFile(group, name, content) {
    const configDir = path.join(this.sshManagerDir, 'config', group);
    const configPath = path.join(configDir, `${name}.conf`);
    
    await fs.ensureDir(configDir);
    await fs.writeFile(configPath, content, 'utf8');
    
    return configPath;
  }

  async readConfigFile(group, name) {
    const configPath = path.join(this.sshManagerDir, 'config', group, `${name}.conf`);
    
    if (await fs.pathExists(configPath)) {
      return await fs.readFile(configPath, 'utf8');
    }
    
    return null;
  }

  async removeConfigFile(group, name) {
    const configPath = path.join(this.sshManagerDir, 'config', group, `${name}.conf`);
    
    if (await fs.pathExists(configPath)) {
      await fs.remove(configPath);
      return true;
    }
    
    return false;
  }

  async listConfigFiles(group = null) {
    const configs = [];
    const configDir = path.join(this.sshManagerDir, 'config');
    
    if (!await fs.pathExists(configDir)) {
      return configs;
    }

    const groups = group ? [group] : await this.listGroups();
    
    for (const groupName of groups) {
      const groupDir = path.join(configDir, groupName);
      
      if (await fs.pathExists(groupDir)) {
        const files = await fs.readdir(groupDir);
        
        for (const file of files) {
          if (file.endsWith('.conf')) {
            const name = path.basename(file, '.conf');
            configs.push({
              group: groupName,
              name: name,
              path: path.join(groupDir, file)
            });
          }
        }
      }
    }
    
    return configs;
  }

  async backupMainConfig() {
    if (await fs.pathExists(this.sshConfigPath)) {
      const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
      const backupPath = path.join(
        this.sshManagerDir, 
        'backups', 
        'config-backups', 
        `ssh-config-${timestamp}.bak`
      );
      
      await fs.copy(this.sshConfigPath, backupPath);
      return backupPath;
    }
    
    return null;
  }

  async readMainConfig() {
    if (await fs.pathExists(this.sshConfigPath)) {
      return await fs.readFile(this.sshConfigPath, 'utf8');
    }
    
    return '';
  }

  async writeMainConfig(content) {
    await fs.ensureDir(this.sshDir);
    await fs.writeFile(this.sshConfigPath, content, 'utf8');
    
    if (process.platform !== 'win32') {
      await fs.chmod(this.sshConfigPath, 0o600);
    }
  }

  async writeTemplate(name, content) {
    const templatePath = path.join(this.sshManagerDir, 'templates', `${name}.conf`);
    await fs.writeFile(templatePath, content, 'utf8');
    return templatePath;
  }

  async readTemplate(name) {
    const templatePath = path.join(this.sshManagerDir, 'templates', `${name}.conf`);
    
    if (await fs.pathExists(templatePath)) {
      return await fs.readFile(templatePath, 'utf8');
    }
    
    return null;
  }

  async listTemplates() {
    const templatesDir = path.join(this.sshManagerDir, 'templates');
    const templates = [];
    
    if (await fs.pathExists(templatesDir)) {
      const files = await fs.readdir(templatesDir);
      
      for (const file of files) {
        if (file.endsWith('.conf')) {
          templates.push(path.basename(file, '.conf'));
        }
      }
    }
    
    return templates;
  }

  getSSHManagerPath() {
    return this.sshManagerDir;
  }

  getSSHConfigPath() {
    return this.sshConfigPath;
  }

  async listGroups() {
    const configDir = path.join(this.sshManagerDir, 'config');
    
    if (!await fs.pathExists(configDir)) {
      return [];
    }

    // Recursively scan for all group directories
    const groups = await this._scanGroupsRecursively(configDir, '');
    return groups.sort();
  }

  /**
   * Recursively scan for group directories
   * @private
   */
  async _scanGroupsRecursively(baseDir, currentPath) {
    const groups = [];
    
    try {
      const items = await fs.readdir(baseDir);
      
      for (const item of items) {
        const itemPath = path.join(baseDir, item);
        const stat = await fs.stat(itemPath);
        
        if (stat.isDirectory()) {
          const groupPath = currentPath ? `${currentPath}/${item}` : item;
          groups.push(groupPath);
          
          // Recursively scan subdirectories
          const subGroups = await this._scanGroupsRecursively(itemPath, groupPath);
          groups.push(...subGroups);
        }
      }
    } catch (error) {
      // If we can't read a directory, skip it but don't fail entirely
      console.warn(`Warning: Could not read directory ${baseDir}:`, error.message);
    }
    
    return groups;
  }

  /**
   * Get groups organized in a hierarchical tree structure
   * @returns {Object} Tree structure with nested groups
   */
  async getGroupsTree() {
    const allGroups = await this.listGroups();
    const tree = {};
    
    for (const groupPath of allGroups) {
      const parsed = this.parseGroupPath(groupPath);
      
      // Build tree structure
      let current = tree;
      for (let i = 0; i < parsed.segments.length; i++) {
        const segment = parsed.segments[i];
        const partialPath = parsed.segments.slice(0, i + 1).join('/');
        
        if (!current[segment]) {
          current[segment] = {
            name: segment,
            fullPath: partialPath,
            children: {},
            connections: []
          };
        }
        current = current[segment].children;
      }
    }
    
    return tree;
  }

  async createGroup(groupPath) {
    // Validate the group path
    if (!this.isValidGroupPath(groupPath)) {
      throw new Error(`Invalid group path: ${groupPath}`);
    }

    const parsed = this.parseGroupPath(groupPath);
    const normalizedPath = parsed.normalized;

    // Create nested directory structure
    const configDir = path.join(this.sshManagerDir, 'config', normalizedPath);
    const keysDir = path.join(this.sshManagerDir, 'keys', normalizedPath);
    
    // Ensure all parent directories exist first
    await fs.ensureDir(configDir);
    await fs.ensureDir(keysDir);
    
    return normalizedPath;
  }

  async renameGroup(oldName, newName) {
    const oldConfigDir = path.join(this.sshManagerDir, 'config', oldName);
    const newConfigDir = path.join(this.sshManagerDir, 'config', newName);
    const oldKeysDir = path.join(this.sshManagerDir, 'keys', oldName);
    const newKeysDir = path.join(this.sshManagerDir, 'keys', newName);

    if (!await fs.pathExists(oldConfigDir)) {
      throw new Error(`Group '${oldName}' does not exist`);
    }

    if (await fs.pathExists(newConfigDir)) {
      throw new Error(`Group '${newName}' already exists`);
    }

    await fs.move(oldConfigDir, newConfigDir);
    
    if (await fs.pathExists(oldKeysDir)) {
      await fs.move(oldKeysDir, newKeysDir);
    } else {
      await fs.ensureDir(newKeysDir);
    }

    return newName;
  }

  async deleteGroup(groupName) {
    const configDir = path.join(this.sshManagerDir, 'config', groupName);
    const keysDir = path.join(this.sshManagerDir, 'keys', groupName);

    if (!await fs.pathExists(configDir)) {
      throw new Error(`Group '${groupName}' does not exist`);
    }

    const files = await fs.readdir(configDir);
    const configFiles = files.filter(file => file.endsWith('.conf'));
    
    if (configFiles.length > 0) {
      throw new Error(`Cannot delete group '${groupName}' because it contains ${configFiles.length} connection(s)`);
    }

    await fs.remove(configDir);
    
    if (await fs.pathExists(keysDir)) {
      await fs.remove(keysDir);
    }

    return groupName;
  }
}

module.exports = FileUtils;

File: src/ssh_manager/backend/file_utils.py
#!/usr/bin/env python3

import os
import json
import shutil
from pathlib import Path
from typing import Dict, List, Optional, Tuple


class FileUtils:
    """File system utilities for SSH Manager."""
    
    def __init__(self):
        self.home_dir = Path.home()
        self.ssh_manager_dir = self.home_dir / 'ssh_manager'
        self.ssh_dir = self.home_dir / '.ssh'
        self.ssh_config_path = self.ssh_dir / 'config'
        self.backup_path = self.ssh_dir / 'config.ssh-manager-backup'
        
    def get_ssh_manager_path(self) -> Path:
        """Get the SSH Manager directory path."""
        return self.ssh_manager_dir
        
    def parse_group_path(self, group_path: str) -> Dict[str, any]:
        """
        Parse a group path into its components.
        
        Args:
            group_path: Group path like 'work/company-a/dev'
            
        Returns:
            Dict with segments, depth, parent, and name
        """
        if not group_path or not isinstance(group_path, str):
            raise ValueError('Group path must be a non-empty string')
            
        # Normalize path - remove leading/trailing slashes, collapse multiple slashes
        normalized = group_path.strip('/')
        if not normalized:
            raise ValueError('Group path cannot be empty after normalization')
            
        segments = normalized.split('/')
        depth = len(segments)
        name = segments[-1]
        parent = '/'.join(segments[:-1]) if depth > 1 else None
        
        return {
            'segments': segments,
            'depth': depth,
            'parent': parent,
            'name': name
        }
        
    def get_group_config_path(self, group_path: str) -> Path:
        """Get the config directory path for a group."""
        parsed = self.parse_group_path(group_path)
        return self.ssh_manager_dir / 'config' / Path(*parsed['segments'])
        
    def get_connection_file_path(self, group_path: str, connection_name: str) -> Path:
        """Get the file path for a specific connection."""
        group_dir = self.get_group_config_path(group_path)
        return group_dir / f"{connection_name}.conf"
        
    def ensure_directory_structure(self):
        """Create the SSH Manager directory structure."""
        directories = [
            self.ssh_manager_dir,
            self.ssh_manager_dir / 'config',
            self.ssh_manager_dir / 'config' / 'personal',
            self.ssh_manager_dir / 'config' / 'work', 
            self.ssh_manager_dir / 'config' / 'projects',
            self.ssh_manager_dir / 'keys',
            self.ssh_manager_dir / 'templates',
            self.ssh_manager_dir / 'backups',
        ]
        
        for directory in directories:
            directory.mkdir(parents=True, exist_ok=True)
            
        # Ensure .ssh directory exists
        self.ssh_dir.mkdir(exist_ok=True)
        
    def create_ssh_config_backup(self) -> Dict[str, any]:
        """Create a backup of the original SSH config."""
        if self.backup_path.exists():
            return {'created': False, 'message': 'Backup already exists'}
            
        if self.ssh_config_path.exists():
            shutil.copy2(self.ssh_config_path, self.backup_path)
            return {'created': True, 'path': str(self.backup_path)}
        else:
            # Create empty backup to indicate SSH Manager was initialized
            self.backup_path.touch()
            return {'created': True, 'path': str(self.backup_path), 'empty': True}
            
    def read_config_file(self, group_path: str, connection_name: str) -> Optional[str]:
        """Read a connection config file."""
        file_path = self.get_connection_file_path(group_path, connection_name)
        if file_path.exists():
            return file_path.read_text(encoding='utf-8')
        return None
        
    def write_config_file(self, group_path: str, connection_name: str, content: str):
        """Write a connection config file."""
        file_path = self.get_connection_file_path(group_path, connection_name)
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_text(content, encoding='utf-8')
        
    def delete_config_file(self, group_path: str, connection_name: str):
        """Delete a connection config file."""
        file_path = self.get_connection_file_path(group_path, connection_name)
        if file_path.exists():
            file_path.unlink()
            
    def list_groups(self) -> List[str]:
        """List all available groups."""
        config_dir = self.ssh_manager_dir / 'config'
        if not config_dir.exists():
            return []
            
        groups = []
        for item in config_dir.rglob('*'):
            if item.is_dir():
                relative_path = item.relative_to(config_dir)
                groups.append(str(relative_path))
                
        return sorted(groups)
        
    def list_connections_in_group(self, group_path: str) -> List[str]:
        """List all connections in a specific group."""
        group_dir = self.get_group_config_path(group_path)
        if not group_dir.exists():
            return []
            
        connections = []
        for file_path in group_dir.glob('*.conf'):
            connection_name = file_path.stem
            connections.append(connection_name)
            
        return sorted(connections)
        
    def get_groups_tree(self) -> Dict[str, any]:
        """Get a hierarchical tree structure of all groups and connections."""
        tree = {}
        
        # Get all groups
        groups = self.list_groups()
        
        for group_path in groups:
            connections = self.list_connections_in_group(group_path)
            parsed = self.parse_group_path(group_path)
            
            # Build nested structure
            current = tree
            for segment in parsed['segments']:
                if segment not in current:
                    current[segment] = {'connections': [], 'subgroups': {}}
                current = current[segment]['subgroups']
                
            # Add connections to the final group
            final_group = tree
            for segment in parsed['segments'][:-1]:
                final_group = final_group[segment]['subgroups']
            if parsed['segments']:
                final_group[parsed['segments'][-1]]['connections'] = connections
                
        return tree
        
    def update_main_ssh_config(self):
        """Update the main SSH config to include SSH Manager configs."""
        # Use recursive glob pattern to match all nested .conf files
        # Note: SSH config supports wildcards including ** for recursive matching in OpenSSH 7.3+
        include_line = f"Include {self.ssh_manager_dir}/config/**/*.conf\n"

        # Read existing config
        existing_config = ""
        if self.ssh_config_path.exists():
            existing_config = self.ssh_config_path.read_text(encoding='utf-8')

        # Check if already included (look for any SSH Manager include)
        if str(self.ssh_manager_dir) in existing_config:
            # Update the include pattern if it's the old single-level pattern
            old_pattern = f"Include {self.ssh_manager_dir}/config/*/*.conf"
            if old_pattern in existing_config:
                existing_config = existing_config.replace(old_pattern, include_line.strip())
                self.ssh_config_path.write_text(existing_config, encoding='utf-8')
            return  # Already included

        # Add include line at the beginning
        new_config = include_line + "\n" + existing_config
        self.ssh_config_path.write_text(new_config, encoding='utf-8')
        
    def revert_to_original_config(self):
        """Revert to the original SSH config."""
        if self.backup_path.exists():
            if self.backup_path.stat().st_size > 0:
                shutil.copy2(self.backup_path, self.ssh_config_path)
            else:
                # Empty backup means no original config existed
                if self.ssh_config_path.exists():
                    self.ssh_config_path.unlink()
        else:
            raise FileNotFoundError("No backup found to revert to")